[FEAT] 멘토링 개설 기능 구현

jye326 opened the pull request:
## Issue Number
closed #164 

## As-Is
- 멘토링 개설 요청 api 필요

## To-Be
- 멘토링 개설 요청 기능 구현

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description
- S3 버킷에 이미지를 실제로 저장하기 위해서는 EC2 인스턴스에 IAM 역할을 할당하거나, AccessKey를 발급해야 합니다. 
- 그러나 현재 규정 상 Access Key 발급이 불가하여, 로컬 환경에서는 테스트할 수 없으며 EC2 인스턴스에 직접 배포하여야만 테스트가 가능합니다.

jye326 committed:
feat: S3 업로더 구현

- 의존성 추가
- S3 설정 클래스 추가
- S3 업로더 클래스 추가

jye326 committed:
refactor: 이미지 타입 수정

- MENTORING -> MENTORING_PROFILE

jye326 committed:
feat: 자격증 dto 추가

jye326 committed:
feat: 멘토링 등록 api 구현

- 카테고리와 멘토링 매핑
- 프로필 이미지 업로드 및 멘토링과 매핑
- 자격증 등록
- 자격증 이미지 업로드 및 자격증과 매핑
- 초기 데이터 수정

jye326 committed:
test: 멘토링 등록 테스트 작성

- 기본 멘토링 정보 등록 테스트
- 프로필 이미지 포함 테스트
- 프로필, 자격증 이미지 포함 테스트
- 트랜잭션 관리를 위해 SpringBootTest로 변경
  - 기존 entityManager 사용 테스트는 유지

ppparkta commented on backend/fittoring/build.gradle:
```diff
@@ -29,14 +29,19 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-actuator'
+    // AWS S3
+    implementation 'software.amazon.awssdk:s3:2.25.24'
+    implementation 'software.amazon.awssdk:auth:2.25.24'
```
p3: auth 의존성은 무슨 역할인가요?

jye326 replied:
> S3에 접근하려면 aws 자격 인증이 필요하고 이를 위한 의존성이 auth입니다.

그런데 놀라운 사실을 발견했습니다.
s3 의존성 안에 이미 auth 의존성이 포함되어있더군요!

화들짝 놀라서 auth는 제거했습니다😀

ppparkta replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 좋네요

ppparkta commented on backend/fittoring/src/main/java/fittoring/config/S3Configuration.java:
```diff
@@ -0,0 +1,19 @@
+    public S3Client s3Client() {
+        return S3Client.builder()
+                .region(Region.AP_NORTHEAST_2)
+                .credentialsProvider(DefaultCredentialsProvider.create())
```
p3: 이름만 봐서는 어떤 자격같은걸 부여하는 코드로 보여져요. 어떤 기능을 하는 코드인지 설명해주실 수 있나요? 
또한, S3Client를 S3라는 외부의 서비스와 통신하기 위한 클라이언트로 이해하면 될까요?

jye326 replied:
> S3Client를 S3라는 외부의 서비스와 통신하기 위한 클라이언트로 이해하면 될까요?

넵⭕

> 이름만 봐서는 어떤 자격같은걸 부여하는 코드로 보여져요. 어떤 기능을 하는 코드인지 설명해주실 수 있나요?

S3에 접근하기 위해 자격 증명이 필요합니다. 자격을 증명하는 방법은 다양한데요. 어떤 방식으로 자격을 증명할 것인지(어떤 제공자 _Provider_ 로부터 자격을 받을 것인지) 설정하는 라인으로 이해하시면 될 것 같습니다.

현재는 `DefaultCredentialsProvider`로 설정해두었고, `DefaultCredentialsProvider`는 아래 순서대로 자격증명을 탐색합니다.

> 
> 1. **환경 변수(Environment Variables)**:
>     - `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`와 같은 환경 변수를 확인.
> 
> 2. **시스템 속성(System Properties)**:
>     - `aws.accessKeyId`, `aws.secretAccessKey`와 같은 시스템 속성을 확인.
> 
> 3. **프로파일 파일(Profile Files)**:
>     - 사용자의 `.aws/credentials` 파일과 **기본 프로파일**을 확인.
>     - 위치: `~/.aws/credentials` (Linux, macOS) 또는 `%USERPROFILE%\.aws\credentials` (Windows).
> 
> 4. **EC2/ECS 자격 증명 서비스**:
>     - AWS에서 실행 중인 애플리케이션의 경우, 제공되는 **IAM 역할**을 통해 자격 증명을 요청.
> 
> 5. **AWS_WEB_IDENTITY_TOKEN_FILE** (OIDC Tokens 사용 시):
>     - OIDC(OpenID Connect)를 사용할 경우, 환경 변수에 설정된 토큰 파일을 검토.
> 
> 

저희 서버에서는 4번을 통해 자격이 증명될 것 같네요!

kimdevv replied:
`DefaultCredentialsProvider.create()`를 해두면 자동으로 저 방식들 중에서 자격을 증명하는 건가요?? 대단하군요...ㄷㄷ

ppparkta replied:
구체적인 설명 감사합니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Certificate.java:
p2: Certificate를 최초로 등록할 때는 곧바로 등록되지 않고 관리자 승인을 거쳐야만 노출되도록 해야하는 것으로 알고있어요. 따라서 Certificate 엔티티 내에 상태가 필요하지 않을까요? (PENDING, APPROVED, REJECTED)

jye326 replied:
`Status verificationStatus;` 필드 추가했습니다.

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/CategoryRepository.java:
```diff
@@ -8,4 +8,6 @@
 
     boolean existsByTitle(String title);
+
+    Category findByTitle(String title);
```
p2: 반환값을 Optional로 처리하는 것이 안정적일 것 같아요!

jye326 replied:
Optional로 변경 후 null이면 예외가 발생하도록 수정했습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Certificate.java:
```diff
@@ -9,21 +9,32 @@
     private Long id;
 
+    @Column(nullable = false)
+    private String type;
```
p2: type은 제한된 네 개의 선택지 중에 선택할 수 있는 방식이니까 enum타입으로 처리하면 편리하고 안정적일 것 같아요!

jye326 replied:
반영했습니다.
```
    LICENSE, // 자격증, 면허
    EDUCATION, // 학력
    AWARD, // 수상 경력
    ETC // 기타
```

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -27,6 +30,9 @@ public class MentoringService {
     private final CategoryRepository categoryRepository;
     private final CategoryMentoringRepository categoryMentoringRepository;
     private final ImageRepository imageRepository;
+    private final S3Uploader s3Uploader;
```
p3: 레포지토리 밑으로 내리면 어떨까요?

jye326 replied:
반영했습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -27,6 +30,9 @@ public class MentoringService {
+    private final S3Uploader s3Uploader;
+    private final CertificateRepository certificateRepository;
+
 
```
p1: 개행 제거해주세요

jye326 replied:
매의 눈..

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -114,12 +120,83 @@ public MentoringResponse getMentoring(final Long id) {
+            throw new S3UploadException(InfraErrorMessage.S3_UPLOAD_ERROR.getMessage());
+        }
+    }
+
```
p1: 여기도요

kimdevv replied:
p3: 이 빈 줄을 지울까요!

kimdevv commented on backend/fittoring/build.gradle:
```diff
@@ -29,14 +29,19 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-actuator'
+    // AWS S3
+    implementation 'software.amazon.awssdk:s3:2.25.24'
+    implementation 'software.amazon.awssdk:auth:2.25.24'
```
p3: 현 시점에서 최신 버전을 사용하는 건 어떨까요?? [이 사이트](https://mvnrepository.com/artifact/software.amazon.awssdk/s3)에서는 2.32.14가 최신 버전이라고 하네요!

jye326 replied:
최신버전으로 변경했습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -114,12 +120,83 @@ public MentoringResponse getMentoring(final Long id) {
+        }
+    }
+
+    private Image saveProfileImage(RegisterMentoringDto dto, Mentoring mentoring) {
+        if (dto.profileImage() == null) {
+            return null;
+        }
+        try {
+            String profileUrl = s3Uploader.upload(dto.profileImage(), "profile-image");
+            Image profile = new Image(profileUrl, ImageType.MENTORING_PROFILE, mentoring.getId());
+            return imageRepository.save(profile);
+        } catch (IOException e) {
+            throw new S3UploadException(InfraErrorMessage.S3_UPLOAD_ERROR.getMessage());
+        }
+    }
```
p2: 이미지 저장은 중복된 코드가 생길 여지가 많아보여요. 설계 의도에 맞게 이미지 서비스로 분리하는 것 어떠신가요?

jye326 replied:
분리했습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/ImageType.java:
```diff
@@ -2,6 +2,7 @@
 public enum ImageType {
 
-    MENTORING,
+    MENTORING_PROFILE,
+    CERTIFICATE,
```
p3: 적절한 의미의 이름이네요 조은데요

kimdevv commented on backend/fittoring/src/main/java/fittoring/config/S3Configuration.java:
```diff
@@ -0,0 +1,19 @@
+    public S3Client s3Client() {
+        return S3Client.builder()
+                .region(Region.AP_NORTHEAST_2)
+                .credentialsProvider(DefaultCredentialsProvider.create())
```
p3: 이 16번 라인은 무슨 기능을 하나요??

jye326 replied:
[답변](https://github.com/woowacourse-teams/2025-Fit-toring/pull/256/files#r2250363204)

kimdevv replied:
DefaultCredentialsProvider.create()를 해두면 자동으로 저 방식들 중에서 자격을 증명하는 건가요?? 대단하군요...ㄷㄷ

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Certificate.java:
```diff
@@ -9,21 +9,32 @@
 @Entity
 public class Certificate {
 
+    @Getter
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
```
p3: 이제 보니 순서가 `@Id`가 맨 아래로 가야 할 것 같은 느낌적인 느낌이네요 나중에 일괄 변경해야할 지도...?

jye326 replied:
가장 정체성에 가까운 어노테이션을 가깝게 붙이는 저희 컨벤션에 따르자면... 

그게 맞네요👍

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -114,12 +120,83 @@ public MentoringResponse getMentoring(final Long id) {
         return MentoringResponse.from(mentoring, categoryTitles, image);
     }
+
+    public MentoringResponse registerMentoring(RegisterMentoringDto dto) {
```
p2: 개인적으로 코드 라인이 길어지니까 이해하는데에 부담이 느껴지네요. 목적에 맞게 서비스를 나눠보는 것 어떠신가요? 조회와 생성/수정/삭제 로직을 분리하는등으로 분리해낼 수 있을 것 같아요!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Certificate.java:
```diff
@@ -9,21 +9,32 @@
     private Long id;
 
+    @Column(nullable = false)
+    private String type;
```
p2: 이 `type` 필드가 자격증, 학력 등을 나타내는 필드인가요?? 그렇다면 입력되는 값을 제한할 수 있게 enum으로 관리하는 건 어떨까요?

jye326 replied:
[답변](https://github.com/woowacourse-teams/2025-Fit-toring/pull/256/files#r2250375033)

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/CategoryRepository.java:
```diff
@@ -8,4 +8,6 @@
 
     boolean existsByTitle(String title);
+
+    Category findByTitle(String title);
```
p2: 바로 `Category`를 반환하게 되면 null을 반환할 수도 있어서 `Optional<Category>`를 반환하는 게 좋을 것 같습니다!

jye326 replied:
[ㄷㅂ](https://github.com/woowacourse-teams/2025-Fit-toring/pull/256/files#r2250370225)

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -114,12 +120,83 @@ public MentoringResponse getMentoring(final Long id) {
+
+    private void categoryMapping(List<String> categoryTitles, Mentoring savedMentoring) {
+        for (String categoryTitle : categoryTitles) {
+            Category category = categoryRepository.findByTitle(categoryTitle);
```
p1: null 체크가 안되고 있네요

jye326 replied:
매의 눈..

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -27,6 +30,9 @@ public class MentoringService {
     private final CategoryRepository categoryRepository;
     private final CategoryMentoringRepository categoryMentoringRepository;
     private final ImageRepository imageRepository;
+    private final S3Uploader s3Uploader;
+    private final CertificateRepository certificateRepository;
```
p3: `certificateRepository`를 위로 올려서 레포지토리끼리 모아두는 건 어떤가요

jye326 replied:
[ㄷ](https://github.com/woowacourse-teams/2025-Fit-toring/pull/256/files#r2250376549)

kimdevv replied:
ㅋㅋ

ppparkta replied:
ㅋㅋㅋ

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -114,12 +120,83 @@ public MentoringResponse getMentoring(final Long id) {
     }
+
+    public MentoringResponse registerMentoring(RegisterMentoringDto dto) {
+        final Mentoring mentoring = new Mentoring(dto.mentorInfo(), dto.mentorInfo(), dto.price(), dto.career(), dto.content(), dto.introduction());
```
p2: `mentorInfo`가 두 번 들어가는 게 맞나요?? `Mentoring.java`가 안 보여서 모르겠네요...!

jye326 replied:
추후에 회원 정보 구현이 완료되면 회원의 이름과 폰 정보가 들어갑니다. 지금은 임시 정보를 넣어놨어요!

```
public record RegisterMentoringDto(
        String mentorInfo,  // 임시 멘토 회원 정보 TODO : 로그인 정보로 멘토 정보 대체하기
        int price,
...
) {
    public static RegisterMentoringDto of(MentoringRequest request, MultipartFile profileImage, List<MultipartFile> certificateImages) {
        return new RegisterMentoringDto(
                "테스트멘토정보",
                request.price(),
...
```

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -114,12 +120,83 @@ public MentoringResponse getMentoring(final Long id) {
+        List<String> categoryTitles = dto.category();
+        categoryMapping(categoryTitles, savedMentoring);
+
+        Image profileImage = saveProfileImage(dto, savedMentoring);
```
p2: saveProfileImage 내부에서 dto.profileImage만 사용하는 것 같은데 여기서는 dto를 곧바로 넘기는 것보다 값을 꺼내서 넣어주는 것이 좋지 않을까요?

jye326 replied:
맞네요! 반영했습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -114,12 +120,83 @@ public MentoringResponse getMentoring(final Long id) {
+        }
+    }
+
+    private Image saveProfileImage(RegisterMentoringDto dto, Mentoring mentoring) {
```
p3: `dto`의 역할은 `profileImage()`를 호출하는 것밖에 없어 보이는데, 매개변수로 `RegisterMentoringDto`를 받지 말고 `dto.profileImage()` 값을 받는 건 어떤가요?

jye326 replied:
[_](https://github.com/woowacourse-teams/2025-Fit-toring/pull/256/files#r2250393422)

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -114,12 +120,83 @@ public MentoringResponse getMentoring(final Long id) {
+        if (validateCertificateRequestData(certificateInfos, certificateImageFiles)) {
+            saveAllCertificates(certificateInfos, certificateImageFiles, savedMentoring);
+        }
+    }
```
p2: 이 부분은 certificate를 담당하는 서비스에서 처리하는 것이 좋지 않을까요?

jye326 replied:
분리했습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -114,12 +120,83 @@ public MentoringResponse getMentoring(final Long id) {
         return MentoringResponse.from(mentoring, categoryTitles, image);
     }
+
+    public MentoringResponse registerMentoring(RegisterMentoringDto dto) {
```
p2: 트랜잭션으로 묶으면 좋을 것 같아요!

jye326 replied:
반영했습니다!

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/business/service/MentoringServiceTest.java:
```diff
@@ -1,47 +1,64 @@
-@Import({DbCleaner.class, MentoringService.class})
-@DataJpaTest
+@Transactional
+@SpringBootTest
```
p3: SpringBootTest를 사용하신 이유가 있나요?

jye326 replied:
- 멘토링 등록은 프로필 이미지와 자격증 등록 트랜잭션이 선행된 후에 각각의 Id 를 조회합니다. 
- 이미지, 자격증 리포지토리가 필요한 큰 작업 단위고 각각 내부의 작은 작업들의 트랜잭션도 관리해주어야 해서 기존 EntityManager 를 사용하기 불편했습니다. 따라서 SpringBootTest 를 추가하였습니다.
- 기존 EntityManager를 사용한 테스트는 변경하지 않았습니다.

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/business/service/MentoringServiceTest.java:
```diff
@@ -312,4 +329,146 @@ void getMentoring2() {
+            MentoringResponse actual = mentoringService.registerMentoring(RegisterMentoringDto.of(request, null, null));
+
+            // then
+            assertAll(
```
p2: softassertions로 변경해주세요

jye326 replied:
변경했습니다!

ppparkta reviewed (commented):
S3를 잘 몰라서 그런지 낯선 코드가 많았네요. 짜느라 정말 고생 많으셨어요! 전반적으로 개행, 테스트 컨벤션 한 번씩만 맞춰주세요. 수고하셨습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/infra/exception/S3UploadException.java:
```diff
@@ -0,0 +1,7 @@
+package fittoring.mentoring.infra.exception;
+
+public class S3UploadException extends RuntimeException {
+    public S3UploadException(String message) {
```
p3: 여길 한 줄 띄울까요?

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/MentoringController.java:
```diff
@@ -32,4 +35,15 @@ public ResponseEntity<MentoringResponse> getMentoring(@PathVariable("mentoringId
+
+    @PostMapping("/mentorings")
+    public ResponseEntity<MentoringResponse> registerMentoring(
+            @RequestPart("data") MentoringRequest request,
+            @RequestPart(value = "image", required = false) MultipartFile profileImage,
+            @RequestPart(value = "certificateImages", required = false) List<MultipartFile> certificateImages
```
p3: 요청을 받을 때 파트를 이렇게 나눠서 받는 건가요??

jye326 replied:
json 형식의 data와 image(multipart)파일, 그리고 이미지파일의 리스트로 나눠받습니다!

kimdevv reviewed (commented):
고생하셨습니다~

jye326 committed:
refactor: 불필요한 awssdk.auth 의존성 제거

jye326 committed:
feat: 자격증 종류 및 인증 상태 enum 구현

jye326 committed:
refactor: Image 작업 서비스 분리

- MentoringService 에서 Image 리포지토리 관련 작업 ImageService 로 분리
- CategoryRepository.findByTitle 반환값 Optional 포장

jye326 committed:
refactor: Certificate 작업 서비스 분리

jye326 reviewed (commented):
리뷰 모두 확인했습니다. 다시 확인해주세요~

jye326 committed:
feat: 멘토링 등록 작업 Transactional 추가

jye326 committed:
refactor: 테스트 형식 변경

- soft assert 로 변경
- 예외 클래스 개행 컨벤션 적용

kimdevv reviewed (approved):
수고하셨습니다~

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/CertificateType.java:
```diff
@@ -0,0 +1,9 @@
+    LICENSE,
+    EDUCATION,
+    AWARD,
+    ETC
```
p3: 컨벤션ㅎ
```java
    ...
    ETC,
    ;
```

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Status.java:
```diff
@@ -0,0 +1,8 @@
+
+    APPROVED,
+    PENDING,
+    REJECTED
```
p3: 아 쪼금 아쉽네요

ppparkta commented on backend/fittoring/src/main/resources/data.sql:
```diff
@@ -135,12 +135,12 @@ VALUES (1, 1),
-       (3, '건강운동관리사'),
-       (4, '홈트레이너 자격증'),
-       (5, 'NSCA CSCS');
+INSERT INTO certificate (mentoring_id, type, title)
+VALUES (1, '자격증','NASM CPT'),
+       (2, '자격증','생활스포츠지도사 2급'),
+       (3, '자격증','건강운동관리사'),
+       (4, '자격증','홈트레이너 자격증'),
+       (5, '자격증','NSCA CSCS');
```
p2: 여기에도 상태가 필요하지 않을까요?

ppparkta reviewed (approved):
수고하셨습니다😄

jye326 committed:
refactor: 컨벤션 반영

ppparkta commented:
좋네요😄 많은 작업이었는데 빠르게 반영해주셔서 감사합니다
